proc/put.go

```diff
@@ -0,0 +1,136 @@
+package proc
+
+import (
+	"github.com/brimsec/zq/ast"
+	"github.com/brimsec/zq/expr"
+	"github.com/brimsec/zq/zbuf"
+	"github.com/brimsec/zq/zcode"
+	"github.com/brimsec/zq/zng"
+)
+
+// The cached information we keep for generating an output record.
+// For a given input descriptor + computed type for the put expression
+// includes the descriptor for output records (outType) plus information
+// about where (position) and how (container) to write the computed value
+// into the output record.
+type descinfo struct {
+	outType   *zng.TypeRecord
+	valType   zng.Type
+	position  int
+	container bool
+}
+
+type Put struct {
+	Base
+	target string
+	eval   expr.ExpressionEvaluator
+	outmap map[int]descinfo
+}
+
+func CompilePutProc(c *Context, parent Proc, node *ast.PutProc) (*Put, error) {
+	eval, err := expr.CompileExpr(node.Expr)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Put{
+		Base:   Base{Context: c, Parent: parent},
+		target: node.Target,
+		eval:   eval,
+		outmap: make(map[int]descinfo),
+	}, nil
+}
+
+func (p *Put) put(in *zng.Record) *zng.Record {
+	val, err := p.eval(in)
+	if err != nil {
+		return in
+	}
+
+	// Figure out the output descriptor.  We cache it in outmap
+	// but if we haven't seen this input descriptor before or if
```

Comment:
What's an example of where the computed type of the expression changes for a given input descriptor? I would have thought that for a given input descriptor, we always get the same output descriptor...

Reply:
A union would be one example -- unions don't work right now, but when they do I don't expect expressions to propagate the union-ness so the expression `x+1` would evaluate to an integer if the concrete value inside `x` is an integer and similarly for float.  Or when we have conditional expressions, `a ? b : c` or `if(a, b, c)` we can't determine at compile time if `b` and `c` are the same type (since that depends on the types in the underlying record).

Reply:
Ah ok. I guess when that time comes we'll probably want to keep those different output types cached somehow rather than recompute them (for unions, there would be one output type per union concrete type). Which of course isn't for now...

Reply:
Yeah.  fwiw, I think the examples above should be uncommon cases so it is reasonable that they aren't handled as efficiently as possible, its more important that the results are correct even for a strange query.

Reply:
True...